TestGelu fails on Grad Diff

### Error
```
380: test_activation_op failed
380:  ..ss..ss..ss..ss..ss..ss..ss..ss...ss..ss.Fss..ss..ss..ss......ssss.....sssss..ss....ssssssss..ss....ss..ssss..ss..ss..ss..ss..ss...ss..ss..ss....ssssss..ss
380: ======================================================================
380: FAIL: test_check_output (test_activation_op.TestGelu)
380: ----------------------------------------------------------------------
380: Traceback (most recent call last):
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/test_activation_op.py", line 56, in test_check_output
380:     self.check_output()
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/op_test.py", line 867, in check_output
380:     check_dygraph)
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/op_test.py", line 834, in check_output_with_place
380:     place, no_check_set=no_check_set, inplace_atol=inplace_atol)
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/op_test.py", line 716, in check_inplace_output_with_place
380:     place, fwd_res, op_desc, inplace_atol=inplace_atol)
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/op_test.py", line 653, in _check_grad_inplace
380:     inplace_atol=inplace_atol)
380:   File "/home/luotao/Paddle/cpu_build/python/paddle/fluid/tests/unittests/op_test.py", line 414, in _compare_expect_and_actual_outputs
380:     + " in class " + self.__class__.__name__ + '\n')
380: AssertionError: Output (X@GRAD) has diff at CPUPlace when using and not using inplace
380: Expect 	lod: {}
380: 	dim: 11, 17
380: 	layout: NCHW
380: 	dtype: float
380: 	data: [0.0324969 0.726696 0.0585252 -0.0132577 0.240575 -0.0303998 0.161463 0.0110685 0.117426 0.0108241 0.289916 -0.0118392 0.000343221 0.0258571 0.00742313 0.140306 0.900668 0.338399 0.0508478 0.00302038 0.210178 0.151127 0.00926887 0.386368 0.0102132 -0.0250177 0.377559 -0.0273923 -0.00864287 -0.00913628 -0.0291934 0.507499 -0.0100213 -0.026525 0.881286 -0.0183183 0.68683 0.690937 0.00450393 -0.0151379 0.00111995 -0.0285519 0.00944021 -0.0191343 0.00395958 -0.0185843 0.420291 0.630156 0.00983283 -0.0273997 0.858117 0.0236381 0.105646 0.0128588 -0.00700907 0.879348 -0.0305546 0.00376284 -0.0157185 0.810201 -0.0304191 0.40073 0.400563 -0.0136204 -0.021126 0.236818 -0.0275397 -0.010712 0.822401 0.136763 0.555896 0.0114044 -0.022969 0.00104584 0.00734837 0.892271 -0.0182961 -0.0305855 -0.0145892 0.00264027 -0.0176496 -0.0272563 0.115599 -0.0249349 -0.030807 0.0128322 -0.0307796 0.0197869 -0.00589312 0.0735134 -0.0264144 0.112631 0.0113269 0.611314 0.0121126 -0.000420736 0.354363 0.00957935 0.228891 0.832072 0.57131 0.227462 0.798018 -0.0242379 0.574287 -0.0307851 0.720939 -0.00376737 0.748282 0.477197 0.508747 2.65738e-05 0.0725443 0.0120923 0.2466 0.0127107 0.00531391 -0.016503 0.579892 -0.0308 0.0239573 0.0458934 -0.0164972 0.0494074 0.150823 -0.0269969 -0.0263068 -0.0184625 0.727809 0.0635955 0.770337 0.0359523 0.000281465 0.0129919 -0.00869936 0.714821 -0.0119533 0.00897189 -0.0280361 -0.0306443 0.535539 0.011293 0.695541 0.180142 0.653978 0.0841829 0.429599 -0.0301255 -0.0306719 -0.0297404 0.309247 -0.030721 -0.0192193 0.741975 0.663798 -0.0306523 0.115844 -0.0239437 -0.013261 -0.0301669 -0.029615 0.597852 0.453089 0.226546 0.80028 -0.025632 -0.0196648 0.000971532 -0.0273041 -0.00321413 0.00439048 0.0792501 -0.0307953 0.0428989 -0.0107674 0.174076 -0.0148395 -0.0245108 0.284196 0.0389162 -0.0300878 0.0261104 0.200669 0.674532 0.0571598 -0.0176833 0.293573]
380:
380: But Got	lod: {}
380: 	dim: 11, 17
380: 	layout: NCHW
380: 	dtype: float
380: 	data: [0.0478344 0.638958 0.0822381 -0.0359259 0.278351 -0.0594018 0.200101 0.0443471 0.152445 0.0430701 0.323364 -0.0197361 0.00111194 0.0385896 0.0272276 0.177634 0.734147 0.365241 0.07236 0.0102095 0.249257 0.189218 0.0354309 0.404643 0.0158216 -0.0450914 0.397549 -0.0505864 -0.0142191 -0.0260376 -0.0632794 0.496372 -0.0282485 -0.0485089 0.724261 -0.0466868 0.615016 0.617521 0.015618 -0.0256201 0.00178088 -0.0627703 0.0362357 -0.048278 0.0135997 -0.04721 0.431399 0.579518 0.0381118 -0.0615679 0.712211 0.0354493 0.139074 0.054581 -0.0114591 0.723263 -0.0601045 0.0128801 -0.0266789 0.686484 -0.0635112 0.416079 0.415947 -0.0367473 -0.0369747 0.274815 -0.05095 -0.0299401 0.693139 0.173797 0.530262 0.0461433 -0.055169 0.00166342 0.0269113 0.729886 -0.0466429 -0.0602601 -0.0389036 0.00886873 -0.0453521 -0.0613992 0.150391 -0.0582664 -0.0620024 0.054415 -0.0615694 0.0299337 -0.0175017 0.100972 -0.0603437 0.147037 0.0457243 0.567322 0.050103 -0.00134759 0.378567 0.0368954 0.267299 0.698364 0.540752 0.265937 0.679765 -0.0433959 0.542762 -0.0626611 0.635552 -0.0115091 0.651578 0.474387 0.497265 8.56824e-05 0.0997809 0.0499853 0.283989 0.0536629 0.0186996 -0.0430012 0.546531 -0.0618451 0.0359028 0.0658727 -0.0429891 0.0704834 0.188895 -0.0610855 -0.0480006 -0.0317955 0.639614 0.0886533 0.66423 0.0525615 0.000448752 0.0554193 -0.0143153 0.631913 -0.0329074 0.034055 -0.0522021 -0.0605783 0.516185 0.0455421 0.620319 0.219348 0.594653 0.113911 0.438589 -0.0636204 -0.0631509 -0.0570174 0.340323 -0.0610642 -0.0484415 0.647915 0.600801 -0.063198 0.150666 -0.0567473 -0.0359333 -0.0636156 -0.0635147 0.558485 0.456454 0.265062 0.681018 -0.0464594 -0.0492903 0.00317788 -0.0614558 -0.00989341 0.015194 0.107967 -0.0617685 0.0619059 -0.0300745 0.213154 -0.0394517 -0.0576276 0.318276 0.0565736 -0.0636226 0.0389465 0.239919 0.607461 0.0804962 -0.04542 0.3266]
380:  in class TestGelu
380:
380:
380: ----------------------------------------------------------------------
380: Ran 156 tests in 3.296s
380:
380: FAILED (failures=1, skipped=79)
380:
1/1 Test #380: test_activation_op ...............***Failed    6.37 sec
```
### How to reproduce
- Machine: `Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz` and `Intel(R) Xeon(R) CPU E5-2620 v3`
   - CI machine works well.
- CMake: `cmake .. -DWITH_GPU=OFF -DWITH_MKLDNN=ON -DCMAKE_BUILD_TYPE=Release -DWITH_TESTING=ON`
- `ctest -R test_activation_op`
- Develop branch, commitid: 826254f664de34eca156c7425a78bb6f2ba28fb3

--- a/paddle/fluid/operators/activation_op.h
+++ b/paddle/fluid/operators/activation_op.h
@@ -340,10 +340,8 @@ struct GeluFunctor : public BaseActivationFunctor<T> {
   }
 };
 
-// gelu_grad(x) = dout * (0.5 * (1 + erf(x / sqrt(2))) + 0.5 * 2 / sqrt(pie) /
-// sqrt(2) * x * exp (-0.5 * sqrt(x)))
-// gelu_grad(x) = dout * (0.5 + 0.5 * erf(x * M_SQRT1_2) + (0.5 * M_2_SQRTPI *
-// M_SQRT1_2) * x * exp (-0.5 * sqrt(x)))
+// gelu_grad(x) = dout * (0.5 * (1 + erf(x / sqrt(2))) + 0.5 * 2 / sqrt(pi) /
+// sqrt(2) * x * exp (-0.5 * x^2))
 template <typename T>
 struct GeluGradFunctor : BaseActivationFunctor<T> {
   template <typename Device, typename X, typename Out, typename dOut,
@@ -353,42 +351,35 @@ struct GeluGradFunctor : BaseActivationFunctor<T> {
     !defined(__OSX__) && !defined(PADDLE_WITH_CUDA)
     auto x_data = x.data();
     auto dx_data = dx.data();
+    auto dout_data = dout.data();
     int n = std::min(x.size(), dx.size());
 
-    std::memset(dx_data, 0, n * sizeof(T));
-
-    // First(dx_data) = erf(x * M_SQRT1_2)
-    math::CBlas<T>::AXPY(n, static_cast<T>(M_SQRT1_2), x_data, 1, dx_data, 1);
-    math::CBlas<T>::VMERF(n, dx_data, dx_data, VML_LA);
-
-    // Second = 0.5 * M_2_SQRTPI * M_SQRT1_2 * x * exp (-0.5 * sqrt(x))
+    auto first = static_cast<T*>(std::malloc(n * sizeof(T)));
+    std::memset(first, 0, n * sizeof(T));
     auto second = static_cast<T*>(std::malloc(n * sizeof(T)));
     std::memset(second, 0, n * sizeof(T));
 
-    math::CBlas<T>::VSQUARE(n, x_data, second);
+    // first = (0.5 * (1 + erf(x / sqrt(2))))
+    math::CBlas<T>::AXPY(n, static_cast<T>(M_SQRT1_2), x_data, 1, first, 1);
+    math::CBlas<T>::VMERF(n, first, first, VML_LA);
     for (int i = 0; i < n; i++) {
-      second[i] *= static_cast<T>(-0.5);
+      first[i] += static_cast<T>(1);
     }
+    math::CBlas<T>::SCAL(n, static_cast<T>(0.5), first, 1);
+
+    // second = (0.5 * 2/sqrt(pi) * 1/sqrt(2) * x * exp(-0.5 * x^2))
+    math::CBlas<T>::VSQUARE(n, x_data, second);
+    math::CBlas<T>::SCAL(n, -static_cast<T>(0.5), second, 1);
     math::CBlas<T>::VEXP(n, second, second);
     math::CBlas<T>::VMUL(n, x_data, second, second);
-    T tmp = static_cast<T>(0.5) * static_cast<T>(M_SQRT1_2) *
-            static_cast<T>(M_2_SQRTPI);
-    for (int i = 0; i < n; i++) {
-      second[i] *= tmp;
-    }
+    math::CBlas<T>::SCAL(n, static_cast<T>(0.5 * M_2_SQRTPI * M_SQRT1_2),
+                         second, 1);
 
-    // Sum = 0.5 * First + Second
-    math::CBlas<T>::AXPY(n, static_cast<T>(0.5), dx_data, 1, second, 1);
+    // dx = dout * (first + second);
+    math::CBlas<T>::VADD(n, first, second, first);
+    math::CBlas<T>::VMUL(n, dout_data, first, dx_data);
 
-    // 0.5 + Sum
-    for (int i = 0; i < n; i++) {
-      second[i] += static_cast<T>(0.5);
-    }
-
-    // * dout
-    auto dout_data = dout.data();
-    math::CBlas<T>::VMUL(n, dout_data, second, dx_data);
-
+    std::free(first);
     std::free(second);
 #else
     auto first = static_cast<T>(0.5) *

